openwisp_monitoring/monitoring/base/models.py

```diff
@@ -95,11 +117,30 @@ def _get_or_create(cls, **kwargs):
         if 'key' in kwargs:
             kwargs['key'] = cls._makekey(kwargs['key'])
         try:
-            lookup_kwargs = kwargs.copy()
+            lookup_kwargs = deepcopy(kwargs)
             if lookup_kwargs.get('name'):
                 del lookup_kwargs['name']
-            metric = cls.objects.get(**lookup_kwargs)
+            extra_tags = lookup_kwargs.pop('extra_tags', {})
+            condition = models.Q(**lookup_kwargs)
+            # extra_tags can contain different fields.
+            # The database lookup will fail if extra_tags
+            # gets updated (e.g. location is added to a device,
+            # organization of a device is changed, etc.)
+            # which results in the creation of new Metric objects.
+            # Hence, only ifname (if present) is used for the
+            # database lookup.
+            if extra_tags.get('ifname'):
+                condition &= models.Q(
+                    extra_tags__contains='"ifname": "{ifname}"'.format(
+                        ifname=extra_tags.get('ifname')
+                    )
+                )
```

Comment:
This is quite hacky.
It would be better to have 2 fields:

- main_tags: we store things like ifname there
- extra_tags: we store things like org id, location id and floorplan id

We could query data only using main tags, object id and content type, so even if some other detail changes, the charts should not end up empty.
I think we still have to store the other details in extra_tags to allow writing the tags properly.

If a new metric is created when the org or location changes, I don't think it's a big deal, especially if we use only the main_tags to query the data.